parcels/kernels/advection.py

```diff
@@ -100,3 +101,161 @@ def AdvectionRK45(particle, fieldset, time):
+        px[1:] = np.where(-px[1:] + px[0] > 180, px[1:]+360, px[1:])
+    if withW:
+        pz = np.array([grid.depth[zi], grid.depth[zi+1]])
+        dz = pz[1] - pz[0]
+    else:
+        dz = 1.
+
+    c1 = fieldset.UV.dist(px[0], px[1], py[0], py[1], grid.mesh, np.dot(i_u.phi2D_lin(xsi, 0.), py))
+    c2 = fieldset.UV.dist(px[1], px[2], py[1], py[2], grid.mesh, np.dot(i_u.phi2D_lin(1., eta), py))
+    c3 = fieldset.UV.dist(px[2], px[3], py[2], py[3], grid.mesh, np.dot(i_u.phi2D_lin(xsi, 1.), py))
+    c4 = fieldset.UV.dist(px[3], px[0], py[3], py[0], grid.mesh, np.dot(i_u.phi2D_lin(0., eta), py))
+    rad = np.pi / 180.
+    deg2m = 1852 * 60.
+    meshJac = (deg2m * deg2m * math.cos(rad * particle.lat)) if grid.mesh == 'spherical' else 1
+    dxdy = fieldset.UV.jacobian(xsi, eta, px, py) * meshJac
+
+    if withW:
+        U0 = direction * fieldset.U.data[ti, zi+1, yi+1, xi] * c4 * dz
+        U1 = direction * fieldset.U.data[ti, zi+1, yi+1, xi+1] * c2 * dz
+        V0 = direction * fieldset.V.data[ti, zi+1, yi, xi+1] * c1 * dz
+        V1 = direction * fieldset.V.data[ti, zi+1, yi+1, xi+1] * c3 * dz
+        if withTime:
+            U0 = U0 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, zi+1, yi+1, xi] * c4 * dz
+            U1 = U1 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, zi+1, yi+1, xi+1] * c2 * dz
+            V0 = V0 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, zi+1, yi, xi+1] * c1 * dz
+            V1 = V1 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, zi+1, yi+1, xi+1] * c3 * dz
+    else:
+        U0 = direction * fieldset.U.data[ti, yi+1, xi] * c4 * dz
+        U1 = direction * fieldset.U.data[ti, yi+1, xi+1] * c2 * dz
+        V0 = direction * fieldset.V.data[ti, yi, xi+1] * c1 * dz
+        V1 = direction * fieldset.V.data[ti, yi+1, xi+1] * c3 * dz
+        if withTime:
+            U0 = U0 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, yi+1, xi] * c4 * dz
+            U1 = U1 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, yi+1, xi+1] * c2 * dz
+            V0 = V0 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, yi, xi+1] * c1 * dz
+            V1 = V1 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, yi+1, xi+1] * c3 * dz
+
+    def compute_ds(F0, F1, r, direction, tol):
+        up = F0 * (1-r) + F1 * r
+        r_target = 1. if direction * up >= 0. else 0.
+        B = F0 - F1
+        delta = - F0
+        B = 0 if abs(B) < tol else B
+
+        if abs(B) > tol:
+            F_r1 = r_target + delta / B
+            F_r0 = r + delta / B
+        else:
+            F_r0, F_r1 = None, None
+
+        if abs(B) < tol and abs(delta) < tol:
+            ds = float('inf')
+        elif B == 0:
+            ds = -(r_target - r) / delta
+        elif F_r1 * F_r0 < tol:
+            ds = float('inf')
+        else:
+            ds = - 1. / B * math.log(F_r1 / F_r0)
+
+        if abs(ds) < tol:
+            ds = float('inf')
```

Comment:
yes, that is numerically more robust that 'equal-zero' operations. Thanks for the change.